File: leetcode/c++/LCR_188_gu_piao_de_zui_da_li_run_lcof.cpp
// https://leetcode.cn/problems/gu-piao-de-zui-da-li-run-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int bestTiming(vector<int>& prices) {
        if (prices.size() < 2)
            return 0;
        int a = 0, b = 0;
        int buy_price = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] < buy_price)
                buy_price = prices[i];
            b = max(prices[i] - buy_price, a);
            a = b;
        }
        return b;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)

File: leetcode/golang/LCR_188_gu_piao_de_zui_da_li_run_lcof.go
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func bestTiming(prices []int) int {
    if len(prices) < 2 {
        return 0
    }
    a, b, buyPrice := 0, 0, prices[0]
    for i := 1; i < len(prices); i++ {
        b = max(prices[i] - buyPrice, a)
        a = b
        if prices[i] < buyPrice {
            buyPrice = prices[i]
        }
    }
    return b
}